core/coreapi/unixfs.go

```diff
@@ -102,7 +103,21 @@ func (api *UnixfsAPI) Add(ctx context.Context, files files.Node, opts ...options
 	bserv := blockservice.New(addblockstore, exch,
 		blockservice.WriteThrough(cfg.Datastore.WriteThrough.WithDefault(config.DefaultWriteThrough)),
 	) // hash security 001
-	dserv := merkledag.NewDAGService(bserv)
+
+	var dserv ipld.DAGService = merkledag.NewDAGService(bserv)
+
+	// wrap the DAGService in a providingDAG service which provides every block written.
+	// note about strategies:
+	//   - "all"/"flat" gets handled directly at the blockstore so no need to provide
+	//   - "roots" gets handled in the pinner
+	//   - "mfs" gets handled in mfs
+	// We need to provide the "pinned" cases only. Added blocks are not
+	// going to be provided by the blockstore (wrong strategy for that),
+	// nor by the pinner (it does not traverse the pinned DAG).
+	if settings.Pin && !settings.OnlyHash &&
+		(api.providingStrategy == "pinned" || api.providingStrategy == "pinned+mfs") {
+		dserv = &providingDagService{dserv, api.provider}
+	}
```

Comment:
Per the comment, we need to ensure we provide blocks while Adding with the "pinned" strategy as this cannot happen in the Pinner (it doesn't traverse the DAG itself). For that, we wrap the DagService.